scenario/context.py

```diff
@@ -117,52 +116,42 @@ def __enter__(self):
         self.ops = ops
         return self
 
-    def run(self) -> Union[ActionOutput, "State"]:
+    def _run(self) -> Union[ActionOutput, "State"]:
         """Emit the event and proceed with charm execution.
 
         This can only be done once.
         """
         if self._emitted:
-            raise AlreadyEmittedError("Can only context.manager.run() once.")
+            raise AlreadyEmittedError("Can only run once.")
         self._emitted = True
 
         # wrap up Runtime.exec() so that we can gather the output state
         self._wrapped_ctx.__exit__(None, None, None)
 
-        self.output = out = self._get_output()
-        return out
+        return self._get_output()
 
     def __exit__(self, exc_type, exc_val, exc_tb):  # noqa: U100
         if not self._emitted:
-            logger.debug("manager not invoked. Doing so implicitly...")
-            self.run()
-
+            logger.debug("event not emitted. Doing so implicitly...")
+            # The output is discarded so we can use the private method.
+            self._run()
 
-class _EventManager(_Manager):
-    if TYPE_CHECKING:  # pragma: no cover
-        output: State  # pyright: ignore[reportIncompatibleVariableOverride]
 
-        def run(self) -> "State":
-            return cast("State", super().run())
+class ManagedEvent(_Manager):
+    charm: CharmBase  # type: ignore
 
-    @property
-    def _runner(self):
-        return self._ctx._run_event  # noqa
+    def run(self) -> "State":
+        return cast("State", super()._run())
 
     def _get_output(self):
         return self._ctx._output_state  # noqa
 
 
-class _ActionManager(_Manager):
-    if TYPE_CHECKING:  # pragma: no cover
-        output: ActionOutput  # pyright: ignore[reportIncompatibleVariableOverride]
-
-        def run(self) -> "ActionOutput":
-            return cast("ActionOutput", super().run())
+class ManagedAction(_Manager):
```

Comment:
shouldn't these types be private?

Reply:
It's a good question. @benhoyt's argument was that it's important to have the API (specifically, the `.run`/`.run_action` and `.charm` attributes) documented, e.g. in automatic reference docs not just in the README (or a how-to).

Users definitely shouldn't be instantiating these objects themselves, and they aren't in the top-level `scenario` namespace as a reflection of that. I also can't think of any cases where you'd need to use them in a type annotation.

They do 'feel' private to me - they are an implementation detail of how the context manager system works, not a part of the grander Scenario design language. I do think reference docs should include mention of the available attributes, though.

@benhoyt what if they were private, but the `Context` docstring had something like:

> Objects can also provide a context manager when tests need to interact with the charm object, for example:
>     ctx = Context(MyCharm)
>     with ctx(ctx.on.start(), State()) as event:
>         # `event.charm` is the `MyCharm` object created to process the event
>         event.run()  # When using `ctx.on.action()`, use `event.run_action()`

The wording can be tidied up a bunch, but generally explaining the small API of the context manager object inside of the `Context` docstring, rather than having the other objects be public. I think the API is small enough (one method, one object) that people wouldn't be looking them up in reference docs much anyway, and would likely start with the `Context` object itself. It also avoids having both `Context.run` and `ManagedEvent.run` in the reference docs.

Reply:
I don't love that the docs would have to explain the `run` and `run_action` methods just as text inside the docstring. I still think they should be broken out as real types in the docs. They are real types the user interacts with (just not their name), so it seems to me they should have first-class documentation.

Reply:
Yeah but IMHO 'we need these types to show up in the docs' is not a good argument to have them public.
Is there no way to have them private and still convince sphinx to document them as if they were public?

Also, how's the code completion on the managers? Will the IDE warn against `ctx(on.install...) as e: e.run_action()`?
If not, we should try using typing.overload.

Reply:
Things are a little different now that there's just a single class and one `run` that always has the same signature.

IDEs/language servers don't really care about the private-ness, so if `Manager` is `_Manager` you'll still get `.run` and `.charm` auto-completion and the docstrings popping up inline.

Sphinx's autodoc functionality can be given a list of private names that should also be included. So we could have `_Manager` and it could still be in the reference docs.

I still don't think that anyone is going to be looking up the `Manager` class or its attributes/methods in reference documentation, so having it there (in addition to what's in `Context`'s docstring) would be for completeness rather than adding actual value.

I also do feel like "there is a class called Manager" is an implementation detail that should be open to change in the future without being considered breaking. The public API is that you get an object that has `run` and `charm` attributes. I think that would most properly be documented like:

```python
class HasRunAndCharm(abc.ABC):
    charm: ops.CharmBase
    @abc.abstractmethod
    def run(): -> State
        ...

class Context:
    def __call__(self, event, state) -> HasRunAndCharm:
        ...
```

But that's noisier than just having the class be public.

My preference is to have the name be private (`_Manager` or `_EventManager`) and if @benhoyt strongly wants it in the reference docs (which I think will be the case even after reading the above :smile:) then adjusting the Sphinx autodoc so that it's included.